### ISSUE TITLE
[CI] You have reached your pull rate limit 

### ISSUE THREAD

**Issue by tedhtchang** [ray-project/ray#39009]
### What happened + What you expected to happen

I recently started getting this message. Possible due to https://docs.docker.com/docker-hub/download-rate-limit/
```
# kubectl apply -f https://raw.githubusercontent.com/ray-project/kuberay/master/ray-operator/config/samples/ray-cluster.mini.yaml
# kubectl get po raycluster-mini-head-9q29l -oyaml
...
    state:
      waiting:
        message: 'rpc error: code = Unknown desc = failed to pull and unpack image
          "docker.io/rayproject/ray:2.6.3": failed to copy: httpReadSeeker: failed
          open: unexpected status code https://registry-1.docker.io/v2/rayproject/ray/manifests/sha256:9db6f33629b743cc0519f17ae9d1f2db986fb1fcc75a56c1b1740fa1fa3ac82c:
          429 Too Many Requests - Server message: toomanyrequests: You have reached
          your pull rate limit. You may increase the limit by authenticating and upgrading:
          https://www.docker.com/increase-rate-limit'
        reason: ErrImagePull
```
We can probably create an alternative repo such as `quay.io/rayproject/ray` similar to this [issue](https://github.com/ray-project/kuberay/issues/1133).
/cc @kevin85421 

### Versions / Dependencies

All Ray images uploaded to docker.io

### Reproduction script

Pulling within the limit may not encounter this problem
kubectl apply -f https://raw.githubusercontent.com/ray-project/kuberay/master/ray-operator/config/samples/ray-cluster.mini.yaml

### Issue Severity

High: It blocks me from completing my task.

**Comment by kevin85421** [comment]
cc @krfricke @aslonnie @can-anyscale

**Comment by aslonnie** [comment]
- me and @can-anyscale are currently not responsible for kuberay's CI.
- CI / production probably should not be using docker hub, especially when launching a cluster or something
- most container registries have rate limiting for unauthenticated calls.

I am also not sure how is this CI related.

**Comment by kevin85421** [comment]
> me and @can-anyscale are currently not responsible for kuberay's CI.

This isn't related to KubeRay CI. @tedhtchang requests to push "Ray images" to both DockerHub and other image registries without rate limitations, such as Quay.

**Comment by aslonnie** [comment]
I think they are trying to pull, not push.

maybe login first with a docker hub token?

or mirror the image to somewhere yourself and modify the yaml file?

not much I can do here really.

**Comment by kevin85421** [comment]
As I understand it, the issue asks Ray CI to push images not only to DockerHub but also to other image registries. This way, they can pull Ray's official images from a registry without rate limitations. Is my understanding correct, @tedhtchang?

**Comment by aslonnie** [comment]
mostly all registries have rate limits.. bandwidth is not free and DoS issues are real..

- ecr rate limit: https://docs.aws.amazon.com/AmazonECR/latest/userguide/service-quotas.html
- gar rate limit: https://cloud.google.com/artifact-registry/quotas

not sure about quay.io , if it does not have rate limit, it should.

also docker hub is not for production use. ray-project/ray images on docker hub is also not for production use. it is for distribution only.

kuberay should not be encoding ray-project/ray in the k8s manifest yaml file. that is just asking for failures when user trying to scale.

self-hosting a registry backed by s3 or filesystem is also pretty simple: https://docs.docker.com/registry/deploying/

**Comment by aslonnie** [comment]
and quay.io seems to have rate limit too, just not very explicit about it:

> only rate limits in the most severe circumstances to maintain service levels (e.g. tens of requests per second from the same IP address).

https://access.redhat.com/articles/5531191

10/s can be pretty easily got hit when a ray cluster is behind a nat gateway.

**Comment by tedhtchang** [comment]
> As I understand it, the issue asks Ray CI to push images not only to DockerHub but also to other image registries. This way, they can pull Ray's official images from a registry without rate limitations. Is my understanding correct, @tedhtchang?

Correct. I was reviewing a KubeRay PR but I failed to pull Kuberay operator and failed to create RayCluster on KinD on a VM. For prod/dev, we are required to use our custom Ray image on Quay.io but I cannot use our custom image for opensource Kuberay development.

For Kuberay image, I could specify the alternative quay.io repo. For Ray image, there is no alternative I can specify. It would be great if you could publish Ray image to an alternative repo as part of you build process.

**Comment by aslonnie** [comment]
we can copy the ray images to places in addition to docker hub on ray release. @kevin85421 I think you can create an account and do the mirroring yourself, or add steps into the ray release process.

but I do not want to push to other places on pipelines (for each commit). we are even thinking about stop publishing the per-commit build.

also, I am just saying that using quay.io is unlikely going to be a reliable solution to the pull rate limit. so maybe rephrase this issue or open a new one if the request is just to add a mirror.

**Comment by tedhtchang** [comment]
>I do not want to push to other places on pipelines (for each commit).

Agree. I mostly used the default ray images comes with Raycluster yaml.
I updated the issue title and lowered the Severity level because I am not getting the rate limit right now.

**Comment by nvtkaszpir** [comment]
Well, currently dockerhub shows 404 to any rayproject repo...

**Comment by aslonnie** [comment]
> Well, currently dockerhub shows 404 to any rayproject repo...

we are looking at it. seems like a UI only issue. images are still pull-able.

**Comment by anyscalesam** [comment]
@nvtkaszpir - are you able to pull?
@tedhtchang - did you run into any rate limit issues the last 2mo?

**Comment by tedhtchang** [comment]
Hey @anyscalesam It was sporadic before. Since last month, I am experiencing this problem every day as long as we are behind company firewall.
